File: engine.js
(function (global) {
  'use strict';

  const BREADTH_WEIGHT = 1.0;
  const MATRIX_EXPECTED_FILLED = 1;
  const MS_PER_YEAR = 365.25 * 24 * 3600 * 1000;

  function parseDate(dateStr) {
    if (!dateStr) return null;
    const d = new Date(dateStr);
    return isNaN(d.getTime()) ? null : d;
  }

  function ageYears(dateStr) {
    const d = parseDate(dateStr);
    return d ? (Date.now() - d.getTime()) / MS_PER_YEAR : null;
  }

  function ageStepIndex(years, buckets) {
    if (years === null) return 0;
    for (let i = 0; i < buckets.length; i++) if (years < buckets[i]) return i + 1;
    return buckets.length + 1;
  }

  function createEngine(DATA) {
    const EFFORT_WEIGHT = typeof DATA.effortWeight === 'number' ? DATA.effortWeight : 0;
    function effortTerm(q) {
      return typeof q.effort === 'number' ? q.effort * EFFORT_WEIGHT : 0;
    }
    const CAUSES = Object.fromEntries(DATA.causes.map((c) => [c.id, c]));
    const ALL_IDS = Object.keys(CAUSES);

    const CAUSE_CHILDREN = (() => {
      const m = {};
      for (const c of DATA.causes) (m[c.parent] ||= []).push(c.id);
      return m;
    })();

    function expandEffects(effects) {
      if (!effects) return effects;
      const out = {};
      for (const [k, v] of Object.entries(effects)) {
        const kids = CAUSE_CHILDREN[k];
        if (!kids) {
          out[k] = v;
          continue;
        }
        for (const id of kids) if (!(id in effects)) out[id] = v;
      }
      return out;
    }

    function expandStepRows(q) {
      const curves = DATA.sliderCurves || {};
      const labels = q.stepLabels || [];
      return q.rows.map((row) => {
        if (row.steps) return row;
        const curve = curves[row.curve] || [];
        const causes = row.causes || [];
        return {
          ...row,
          steps: labels.map((label, i) => ({
            label,
            effects: i === 0 ? {} : Object.fromEntries(causes.map((id) => [id, curve[i - 1] ?? 0])),
          })),
        };
      });
    }

    const QUESTIONS = DATA.questions
      .map((q) => {
        const baseType = q.type || (q.multiselect ? 'multi' : 'options');
        const next = { ...q, type: baseType };
        if (next.type === 'matrix') {
          next.colMul = Object.fromEntries(next.columns.map((c) => [c.id, c.mult]));
          next.rows = next.rows.map((r) => ({ ...r, effects: expandEffects(r.effects) }));
        } else if (next.type === 'ages') {
          next.rows = expandStepRows(next).map((r) => ({
            ...r,
            steps: r.steps.map((s) => ({ ...s, effects: expandEffects(s.effects) })),
          }));
        } else {
          // 'options' and 'multi' share the same option shape
          next.options = next.options.map((o) => ({ ...o, effects: expandEffects(o.effects) }));
        }
        return next;
      })
      .sort((a, b) => (a.stage ?? Infinity) - (b.stage ?? Infinity));

    const MAIN_QUESTIONS = QUESTIONS.filter((q) => !q.optional);
    const OPTIONAL_QUESTIONS = QUESTIONS.filter((q) => q.optional);
    const AGES_Q = QUESTIONS.find((q) => q.type === 'ages') || null;
    const Q_BY_ID = Object.fromEntries(QUESTIONS.map((q) => [q.id, q]));

    const TYPE_HANDLERS = {
      options: {
        score(q, ans, s) {
          const opt = q.options[ans];
          if (!opt) return;
          for (const [id, delta] of Object.entries(opt.effects)) {
            s[id] = (s[id] || 0) + delta;
          }
        },
        causeTerms(q, ids) {
          let isolation = 0;
          let breadth = 0;
          for (const causeId of ids) {
            const deltas = q.options.map((o) => o.effects[causeId] || 0);
            const spread = Math.max(...deltas) - Math.min(...deltas);
            if (spread > 0) breadth++;
            isolation += spread;
          }
          return { isolation, breadth: BREADTH_WEIGHT * breadth };
        },
        isAnswered() {
          return true;
        },
      },
      multi: {
        score(q, ans, s) {
          if (!Array.isArray(ans)) return;
          for (const i of ans) {
            const opt = q.options[i];
            if (!opt) continue;
            for (const [id, delta] of Object.entries(opt.effects)) {
              s[id] = (s[id] || 0) + delta;
            }
          }
        },
        // Multiselect isolation = sum of |effects| across options (max subset
        // sum minus min subset sum = sum of positives + |sum of negatives|).
        causeTerms(q, ids) {
          let isolation = 0;
          let breadth = 0;
          for (const causeId of ids) {
            let sumAbs = 0;
            for (const opt of q.options) sumAbs += Math.abs(opt.effects[causeId] || 0);
            if (sumAbs > 0) breadth++;
            isolation += sumAbs;
          }
          return { isolation, breadth: BREADTH_WEIGHT * breadth };
        },
        isAnswered(_q, ans) {
          return Array.isArray(ans) && ans.length > 0;
        },
      },
      matrix: {
        score(q, ans, s) {
          const colMul = q.colMul;
          for (const row of q.rows) {
            const m = colMul[ans[row.id] || 'no'] || 0;
            if (m === 0) continue;
            for (const [id, delta] of Object.entries(row.effects)) {
              s[id] = (s[id] || 0) + delta * m;
            }
          }
        },
        causeTerms(q, ids) {
          const mults = q.columns.map((c) => c.mult);
          const multSpread = Math.max(...mults) - Math.min(...mults);
          const p = q.rows.length > 0 ? MATRIX_EXPECTED_FILLED / q.rows.length : 0;
          let isolation = 0;
          const rowsAffecting = {};
          for (const row of q.rows) {
            for (const causeId of ids) {
              const e = Math.abs(row.effects[causeId] || 0);
              if (e > 0) {
                isolation += e * multSpread * p;
                rowsAffecting[causeId] = (rowsAffecting[causeId] || 0) + 1;
              }
            }
          }
          let breadth = 0;
          for (const k of Object.values(rowsAffecting)) {
            breadth += 1 - Math.pow(1 - p, k);
          }
          return { isolation, breadth: BREADTH_WEIGHT * breadth };
        },
        isAnswered(_q, ans) {
          return !!ans && typeof ans === 'object';
        },
      },
      ages: {
        score(q, ans, s) {
          for (const row of q.rows) {
            const idx = ans[row.id];
            if (idx === undefined || idx === null) continue;
            const step = row.steps[idx];
            if (!step) continue;
            for (const [id, delta] of Object.entries(step.effects)) {
              s[id] = (s[id] || 0) + delta;
            }
          }
        },
        causeTerms(q, ids) {
          let isolation = 0;
          const affected = new Set();
          for (const row of q.rows) {
            for (const causeId of ids) {
              const deltas = row.steps.map((st) => st.effects[causeId] || 0);
              const spread = Math.max(...deltas) - Math.min(...deltas);
              if (spread > 0) {
                isolation += spread;
                affected.add(causeId);
              }
            }
          }
          return { isolation, breadth: BREADTH_WEIGHT * affected.size };
        },
        isAnswered(_q, ans) {
          return !!ans && typeof ans === 'object';
        },
      },
    };

    const STAGES = DATA.stages.map((s) => s.id);
    const STAGE_LABELS = Object.fromEntries(DATA.stages.map((s) => [s.id, s.label]));
    const STAGE_QS = Object.fromEntries(
      STAGES.map((s) => [s, QUESTIONS.filter((q) => q.stage === s)])
    );

    function isAnswered(qid, ans) {
      if (ans === undefined || ans === null) return false;
      const q = Q_BY_ID[qid];
      if (!q) return false;
      return TYPE_HANDLERS[q.type].isAnswered(q, ans);
    }

    function isCompleted(qid, answers, skipped) {
      return isAnswered(qid, answers && answers[qid]) || !!(skipped && skipped[qid]);
    }

    function rank(answers) {
      const s = {};
      for (const id of ALL_IDS) s[id] = CAUSES[id].baseline;
      for (const q of QUESTIONS) {
        const ans = answers && answers[q.id];
        if (ans === undefined || ans === null) continue;
        TYPE_HANDLERS[q.type].score(q, ans, s);
      }
      let posTotal = 0;
      for (const id of ALL_IDS) posTotal += Math.max(0, s[id]);
      return ALL_IDS.map((id) => ({
        id,
        score: s[id],
        pct: posTotal > 0 ? (Math.max(0, s[id]) / posTotal) * 100 : 0,
      })).sort((a, b) => b.score - a.score);
    }

    function contendingIds(answers) {
      const ranked = rank(answers);
      if (ranked.length === 0) return [];
      const leader = ranked[0].pct;
      const cutoff = Math.max(2, leader * 0.3);
      const ids = ranked.filter((r) => r.pct >= cutoff).map((r) => r.id);
      return ids.length >= 3 ? ids : ranked.slice(0, 3).map((r) => r.id);
    }

    function requiresMet(q, answers) {
      if (!q.requires) return true;
      for (const [depQid, allowed] of Object.entries(q.requires)) {
        const ans = answers && answers[depQid];
        if (typeof ans !== 'number') return false;
        if (!allowed.includes(ans)) return false;
      }
      return true;
    }

    function discriminators(answers, skipped) {
      const ids = contendingIds(answers);
      const map = {};
      let max = 0;
      for (const q of QUESTIONS) {
        if (isCompleted(q.id, answers, skipped)) continue;
        if (!requiresMet(q, answers)) continue;
        const { isolation, breadth } = TYPE_HANDLERS[q.type].causeTerms(q, ids);
        // Gate: only surface a question that actually discriminates a contending
        // cause. Without this, the constant effort term keeps every unanswered
        // question on the list, so a cheap-but-irrelevant question can outrank a
        // costly one that is the only thing bearing on the remaining causes.
        if (isolation + breadth <= 0) continue;
        const D = isolation + breadth + effortTerm(q);
        map[q.id] = D;
        if (D > max) max = D;
      }
      return { map, max };
    }

    function recommendations(answers, skipped) {
      const { map } = discriminators(answers, skipped);
      return Object.entries(map)
        .filter(([, D]) => D > 0)
        .map(([qid, D]) => ({ q: Q_BY_ID[qid], D }))
        .sort((a, b) => b.D - a.D);
    }

    function relevancyLevel(qid, answers, skipped) {
      const { map, max } = discriminators(answers, skipped);
      const D = map[qid];
      if (D === undefined || D <= 0 || max <= 0) return null;
      const ratio = D / max;
      if (ratio >= 2 / 3) return 'high';
      if (ratio >= 1 / 3) return 'mid';
      return 'low';
    }

    function stageProgress(answers, skipped) {
      const out = {};
      for (const s of STAGES) {
        const qs = STAGE_QS[s];
        out[s] = {
          total: qs.length,
          answered: qs.reduce((n, q) => n + (isCompleted(q.id, answers, skipped) ? 1 : 0), 0),
        };
      }
      return out;
    }

    return {
      CAUSES,
      ALL_IDS,
      QUESTIONS,
      MAIN_QUESTIONS,
      OPTIONAL_QUESTIONS,
      AGES_Q,
      Q_BY_ID,
      STAGES,
      STAGE_LABELS,
      STAGE_QS,
      TYPE_HANDLERS,
      isAnswered,
      isCompleted,
      rank,
      contendingIds,
      discriminators,
      recommendations,
      relevancyLevel,
      stageProgress,
    };
  }

  global.createEngine = createEngine;
  global.parseDate = parseDate;
  global.ageYears = ageYears;
  global.ageStepIndex = ageStepIndex;
})(typeof window !== 'undefined' ? window : globalThis);

File: tools/engine.py
from __future__ import annotations

from typing import Any


class Engine:
    def __init__(self, data: dict[str, Any]):
        self._data = data
        w = data.get("weights", {})
        self._effort_weight: float = float(w.get("effort", 0.0))
        self._breadth_weight: float = float(w.get("breadth", 1.0))
        self._ease_weight: float = float(w.get("ease", 0.5))
        self._matrix_expected_filled: float = float(w.get("matrixExpectedFilled", 1))
        self.failure_modes: dict[str, dict] = {c["id"]: c for c in data["failure_modes"]}
        self.all_fcodes: list[str] = list(self.failure_modes.keys())

        self._mode_children: dict[str, list[str]] = {}
        for c in data["failure_modes"]:
            self._mode_children.setdefault(c["parent"], []).append(c["id"])

        self.questions: list[dict] = self._build_questions()
        self.main_questions = [q for q in self.questions if not q.get("optional")]
        self.q_by_id = {q["id"]: q for q in self.questions}

    def _expand_effects(self, effects: dict | None) -> dict | None:
        if not effects:
            return effects
        out: dict[str, float] = {}
        for k, v in effects.items():
            kids = self._mode_children.get(k)
            if not kids:
                out[k] = v
                continue
            for fcode in kids:
                if fcode not in effects:
                    out[fcode] = v
        return out

    def _expand_step_rows(self, q: dict) -> list[dict]:
        curves = self._data.get("sliderCurves", {})
        labels = q.get("stepLabels", [])
        result = []
        for row in q["rows"]:
            if row.get("steps"):
                result.append(row)
                continue
            curve = curves.get(row.get("curve"), [])
            failure_modes = row.get("failure_modes", [])
            steps = []
            for i, label in enumerate(labels):
                if i == 0:
                    effects: dict[str, float] = {}
                else:
                    val = curve[i - 1] if i - 1 < len(curve) else 0
                    effects = {fcode: val for fcode in failure_modes}
                steps.append({"label": label, "effects": effects})
            result.append({**row, "steps": steps})
        return result

    def _build_questions(self) -> list[dict]:
        result = []
        for q in self._data["questions"]:
            base_type = q.get("type") or ("multi" if q.get("multiselect") else "options")
            nxt = {**q, "type": base_type}
            if nxt["type"] == "matrix":
                nxt["colMul"] = {c["id"]: c["multiplier"] for c in nxt["columns"]}
                nxt["rows"] = [
                    {**r, "effects": self._expand_effects(r.get("effects"))} for r in nxt["rows"]
                ]
            elif nxt["type"] == "ages":
                expanded = self._expand_step_rows(nxt)
                nxt["rows"] = [
                    {
                        **r,
                        "steps": [
                            {**s, "effects": self._expand_effects(s.get("effects"))}
                            for s in r["steps"]
                        ],
                    }
                    for r in expanded
                ]
            else:
                nxt["options"] = [
                    {**o, "effects": self._expand_effects(o.get("effects"))} for o in nxt["options"]
                ]
            result.append(nxt)
        result.sort(key=lambda q: q.get("stage") if q.get("stage") is not None else float("inf"))
        return result

    def _score(self, q: dict, ans: Any, s: dict[str, float]) -> None:
        t = q["type"]
        if t == "options":
            if not isinstance(ans, int) or isinstance(ans, bool) or ans < 0 or ans >= len(q["options"]):
                return
            for fcode, delta in q["options"][ans]["effects"].items():
                s[fcode] = s.get(fcode, 0) + delta
        elif t == "multi":
            if not isinstance(ans, list):
                return
            for i in ans:
                if not isinstance(i, int) or isinstance(i, bool) or i < 0 or i >= len(q["options"]):
                    continue
                for fcode, delta in q["options"][i]["effects"].items():
                    s[fcode] = s.get(fcode, 0) + delta
        elif t == "matrix":
            if not isinstance(ans, dict):
                return
            col_mul = q["colMul"]
            for row in q["rows"]:
                m = col_mul.get(ans.get(row["id"], "no"), 0)
                if m == 0:
                    continue
                for fcode, delta in row["effects"].items():
                    s[fcode] = s.get(fcode, 0) + delta * m
        elif t == "ages":
            if not isinstance(ans, dict):
                return
            for row in q["rows"]:
                idx = ans.get(row["id"])
                if idx is None or not isinstance(idx, int) or idx < 0 or idx >= len(row["steps"]):
                    continue
                for fcode, delta in row["steps"][idx]["effects"].items():
                    s[fcode] = s.get(fcode, 0) + delta

    def _failure_mode_terms(self, q: dict, ids: list[str]) -> tuple[float, float]:
        t = q["type"]
        if t == "options":
            iso = 0.0
            breadth = 0
            for fcode in ids:
                deltas = [o["effects"].get(fcode, 0) for o in q["options"]]
                spread = max(deltas) - min(deltas)
                if spread > 0:
                    breadth += 1
                iso += spread
            return iso, self._breadth_weight * breadth
        if t == "multi":
            iso = 0.0
            breadth = 0
            for fcode in ids:
                sum_abs = sum(abs(o["effects"].get(fcode, 0)) for o in q["options"])
                if sum_abs > 0:
                    breadth += 1
                iso += sum_abs
            return iso, self._breadth_weight * breadth
        if t == "matrix":
            mults = [c["multiplier"] for c in q["columns"]]
            mult_spread = max(mults) - min(mults)
            n_rows = len(q["rows"])
            p = (self._matrix_expected_filled / n_rows) if n_rows > 0 else 0
            iso = 0.0
            rows_affecting: dict[str, int] = {}
            for row in q["rows"]:
                for fcode in ids:
                    e = abs(row["effects"].get(fcode, 0))
                    if e > 0:
                        iso += e * mult_spread * p
                        rows_affecting[fcode] = rows_affecting.get(fcode, 0) + 1
            breadth = sum(1 - (1 - p) ** k for k in rows_affecting.values())
            return iso, self._breadth_weight * breadth
        if t == "ages":
            iso = 0.0
            affected = set()
            for row in q["rows"]:
                for fcode in ids:
                    deltas = [st["effects"].get(fcode, 0) for st in row["steps"]]
                    spread = max(deltas) - min(deltas)
                    if spread > 0:
                        iso += spread
                        affected.add(fcode)
            return iso, self._breadth_weight * len(affected)
        return 0.0, 0.0

    def discriminator_terms(self, q: dict, ids: list[str]) -> dict[str, float]:
        iso, breadth = self._failure_mode_terms(q, ids)
        effort = self._effort_term(q)
        return {"isolation": iso, "breadth": breadth, "effort": effort}

    def _ans_is_present(self, q: dict, ans: Any) -> bool:
        t = q["type"]
        if t == "options":
            return True
        if t == "multi":
            return isinstance(ans, list) and len(ans) > 0
        if t == "matrix":
            return isinstance(ans, dict)
        if t == "ages":
            return isinstance(ans, dict)
        return False

    def is_answered(self, question_id: str, ans: Any) -> bool:
        if ans is None:
            return False
        q = self.q_by_id.get(question_id)
        if not q:
            return False
        return self._ans_is_present(q, ans)

    def is_completed(self, question_id: str, answers: dict | None = None, skipped: dict | None = None) -> bool:
        answers = answers or {}
        skipped = skipped or {}
        return self.is_answered(question_id, answers.get(question_id)) or bool(skipped.get(question_id))

    def rank(self, answers: dict | None = None) -> list[dict]:
        answers = answers or {}
        s: dict[str, float] = {fcode: self.failure_modes[fcode]["baseline"] for fcode in self.all_fcodes}
        for q in self.questions:
            ans = answers.get(q["id"])
            if ans is None:
                continue
            self._score(q, ans, s)
        pos_total = sum(max(0, v) for v in s.values())
        result = [
            {
                "id": fcode,
                "score": s[fcode],
                "percent": (max(0, s[fcode]) / pos_total * 100) if pos_total > 0 else 0,
            }
            for fcode in self.all_fcodes
        ]
        result.sort(key=lambda r: -r["score"])
        return result

    def contending_ids(self, answers: dict | None = None) -> list[str]:
        ranked = self.rank(answers)
        if not ranked:
            return []
        leader = ranked[0]["percent"]
        cutoff = max(2, leader * 0.3)
        ids = [r["id"] for r in ranked if r["percent"] >= cutoff]
        if len(ids) >= 3:
            return ids
        return [r["id"] for r in ranked[:3]]

    def _effort_term(self, q: dict) -> float:
        lvl = q.get("effort")
        return float(lvl) * self._effort_weight if isinstance(lvl, (int, float)) else 0.0

    def _requires_met(self, q: dict, answers: dict | None) -> bool:
        req = q.get("requires")
        if not req:
            return True
        for dep_qid, allowed in req.items():
            ans = (answers or {}).get(dep_qid)
            if not isinstance(ans, int) or isinstance(ans, bool):
                return False
            if ans not in allowed:
                return False
        return True

    def discriminators(self, answers: dict | None = None, skipped: dict | None = None) -> dict:
        ids = self.contending_ids(answers)
        cands: list[tuple[str, float, float]] = []
        max_effort = 0.0
        for q in self.questions:
            if self.is_completed(q["id"], answers, skipped):
                continue
            if not self._requires_met(q, answers):
                continue
            t = self.discriminator_terms(q, ids)
            info = t["isolation"] + t["breadth"]
            if info <= 0:
                continue
            effort = t["effort"]
            if effort > max_effort:
                max_effort = effort
            cands.append((q["id"], info, effort))
        m: dict[str, float] = {}
        mx = 0.0
        for question_id, info, effort in cands:
            # ease is a bounded tie-breaker; must not outweigh separation (info).
            ease = (effort / max_effort) if max_effort > 0 else 0.0
            D = info * (1 + self._ease_weight * ease)
            m[question_id] = D
            if D > mx:
                mx = D
        return {"map": m, "max": mx}

    def recommendations(self, answers: dict | None = None, skipped: dict | None = None) -> list[dict]:
        d = self.discriminators(answers, skipped)
        items = [
            {"q": self.q_by_id[question_id], "D": D}
            for question_id, D in d["map"].items()
            if D > 0
        ]
        items.sort(key=lambda x: -x["D"])
        return items
